packages/block/dom.ts

```diff
@@ -0,0 +1,172 @@
+) => {
+  const event = name.toLowerCase().slice(2);
+  if (!document[EVENTS_REGISTRY]) document[EVENTS_REGISTRY] = new Set();
+  if (!setHas$.call(document[EVENTS_REGISTRY], event)) {
+    // createEventListener uses a synthetic event handler to capture events
+    // https://betterprogramming.pub/whats-the-difference-between-synthetic-react-events-and-javascript-events-ba7dbc742294
+    addEventListener$.call(
+      document,
+      event,
+      (nativeEventObject: Event) => {
+        let el = nativeEventObject.target as Node | null;
+        // Bubble up the DOM tree to find all event listeners
+        while (el) {
+          const pool = el[EVENT_LISTENERS_POOL];
+          if (pool) {
+            const listeners = pool[event] as Record<string, any> | undefined;
+            if (listeners) {
+              const handlers = Object.values(listeners);
+              let returnFalse = false;
+              for (let i = 0, j = handlers.length; i < j; ++i) {
+                if (handlers[i](nativeEventObject) === false) {
+                  returnFalse = true;
+                }
+              }
+              if (returnFalse) return;
+            }
+          }
+          el = el.parentNode;
+        }
+      },
+      {
+        capture: false,
+      },
+    );
+    document[EVENTS_REGISTRY].add(event);
+  }
+  const pointer = listenerPointer++;
+  if (mapHas$.call(eventCache, value)) return mapGet$.call(eventCache, value);
+  const patch = (newValue?: EventListener) => {
+    if (!el[EVENT_LISTENERS_POOL]) el[EVENT_LISTENERS_POOL] = {};
+    const pool = el[EVENT_LISTENERS_POOL];
+    if (!pool[event]) pool[event] = {};
+    pool[event][pointer] = (e: Event) => {
+      (newValue ?? value)?.(e);
+    };
+  };
+  patch(value);
+  mapSet$.call(eventCache, value, patch);
+  return patch;
+};
+
+export const insertText = (
+  el: HTMLElement,
+  value: any,
+  index: number,
+): Text => {
+  const node = document.createTextNode(value);
+  const childNodes = childNodes$.call(el);
+  if (hasChildNodes$.call(el) && index < childNodes.length) {
+    insertBefore$.call(el, node, childNodes[index]);
+  } else {
+    appendChild$.call(el, node);
+  }
+  return node;
+};
+
+export const setText = (el: HTMLElement, value: string, index: number) => {
+  setTextContent$.call(childNodes$.call(el)[index], value);
+};
+
+export const setAttribute = (
+  el: HTMLElement,
+  name: string,
+  value?: string | boolean | null,
+): void => {
+  const isValueNully = value === undefined || value === null;
+  value = isValueNully ? '' : value;
+  if (name === 'style') {
+    if (typeof value === 'string') {
+      el.style.cssText = value;
+    } else if (name.startsWith('-')) {
+      el.style.setProperty(name, String(value));
+    } else if (isValueNully) {
+      el.style[name] = '';
+    } else if (typeof value !== 'number' || IS_NON_DIMENSIONAL.test(name)) {
+      el.style[name] = value;
+    } else {
+      el.style[name] = `${String(value)}px`;
+    }
```

Comment:
IMHO, the number prop should be more common than CSS variable prop. Let's lift `typeof value === 'number'` above.